# Q1. Install MLflow

In [1]:
!mlflow --version

mlflow, version 1.26.0


# Q2. Download and preprocess the data

In [2]:
!python preprocess_data.py --raw_data_path ./data/ --dest_path ./output

4 files were created in the output folder: dv.pkl, test.pkl, train.pkl, valid.pkl

# Q3. Train a model with autolog

In [5]:
!python train.py --data_path ./output

2022/05/29 12:48:53 INFO mlflow.tracking.fluent: Experiment with name 'homework2-exp' does not exist. Creating a new experiment.
2022/05/29 12:48:53 WARNING mlflow.utils.autologging_utils: You are using an unsupported version of sklearn. If you encounter errors during autologging, try upgrading / downgrading sklearn to a supported version, or try upgrading MLflow.


In [6]:
!mlflow ui --backend-store-uri sqlite:///mlflow.db

[2022-05-29 12:51:30 +0200] [41240] [INFO] Starting gunicorn 20.1.0
[2022-05-29 12:51:30 +0200] [41240] [INFO] Listening at: http://127.0.0.1:5000 (41240)
[2022-05-29 12:51:30 +0200] [41240] [INFO] Using worker: sync
[2022-05-29 12:51:30 +0200] [41241] [INFO] Booting worker with pid: 41241
^C
[2022-05-29 12:57:24 +0200] [41240] [INFO] Handling signal: int
[2022-05-29 12:57:24 +0200] [41241] [INFO] Worker exiting (pid: 41241)


| Name        | Value       |
| ----------- | ----------- |
|bootstrap	  |True         |
|ccp_alpha	  |0.0          |
|criterion	  |squared_error|
|max_depth	  |10           | 
|max_features |1.0          |
|max_leaf_nodes	|None       ||
|max_samples	|None       |        |
|min_impurity_decrease	|0.0|               |
|min_samples_leaf	|1      |          |
|min_samples_split	|2      |          |
|min_weight_fraction_leaf	|0.0|              |
|n_estimators	|100        |          |
|n_jobs	|None               |
|oob_score	|False          |
|random_state	|0          |
|verbose	|0              |
|warm_start	|False          |

17 parameters are automatically logged

# Q4. Launch the tracking server locally

!mlflow ui --backend-store-uri sqlite:///mlflow.db --default-artifact-root ./artifacts

default-artifact-root is required to properly setup the server

# Q5. Tune the hyperparameters of the model

In [7]:
!python hpo.py

2022/05/29 13:24:54 INFO mlflow.tracking.fluent: Experiment with name 'random-forest-hyperopt' does not exist. Creating a new experiment.
100%|█████████| 50/50 [04:56<00:00,  5.92s/trial, best loss: 6.6284257482044735]


![hpo](images/hpo_results.png)

The lowest rmse is 6.628.

# Q6. Promote the best model to the model registry

In [12]:
!python register_model.py

2022/05/29 23:19:42 WARNING mlflow.utils.autologging_utils: You are using an unsupported version of sklearn. If you encounter errors during autologging, try upgrading / downgrading sklearn to a supported version, or try upgrading MLflow.
run id: 0067710766b34c83aea513e01df70e5f, rmse: 6.5489
run id: 124c5764258b4f638cd286487f89b3dc, rmse: 6.5491
run id: 77cc48e796774c75a8a6e1b2dbf8a5f2, rmse: 6.5491
run id: 84588225f86a40f49800b2197718bfee, rmse: 6.5498
run id: c10e2dc058784a5f91ea58cb9da1f282, rmse: 6.5479
Successfully registered model 'rf_model'.
2022/05/29 23:20:45 INFO mlflow.tracking._model_registry.client: Waiting up to 300 seconds for model version to finish creation.                     Model name: rf_model, version 1
Created version '1' of model 'rf_model'.


the best model has a rmse of 6.55 for the test data set